lxd/storage/drivers/interface.go

```diff
@@ -0,0 +1,52 @@
+package drivers
+
+import (
+	"github.com/lxc/lxd/lxd/migration"
+	"github.com/lxc/lxd/lxd/operations"
+	"github.com/lxc/lxd/shared/api"
+)
+
+// VolumeType represents a storage volume type.
+type VolumeType string
+
+// VolumeTypeCustom represents a custom storage volume.
+const VolumeTypeCustom = VolumeType("custom")
+
+// VolumeTypeContainer represents a container storage volume.
+const VolumeTypeContainer = VolumeType("containers")
+
+// VolumeTypeImage represents an image storage volume.
+const VolumeTypeImage = VolumeType("images")
+
+// VolumeTypeVM represents a virtual-machine storage volume.
+const VolumeTypeVM = VolumeType("virtual-machines")
+
+// driver is the extended internal interface
+type driver interface {
+	Driver
+
+	create(dbPool *api.StoragePool) error
+}
+
+// Driver repreents a low-level storage driver.
+type Driver interface {
+	// Internal
+	Name() string
+	Version() string
+
+	// Pool
+	Delete(op *operations.Operation) error
+	Mount() (bool, error)
+	Unmount() (bool, error)
+	GetResources() (*api.ResourcesStoragePool, error)
+
+	// Volumes
+	CreateVolume(volType VolumeType, volID int64, volName string, op *operations.Operation) error
```

Comment:
Hmm, so based on past discussions, I think our best bet is to change `CreateVolume` and `CopyVolume` to be:
```go
CreateVolume(volType VolumeType, contentType contentType, name string, filler func(path string) error, op *operations.Operation) error
CreateVolumeFromCopy(volType VolumeType, name string, srcType VolumeType, srcName string, snapshots bool, op *operations.Operations) error
```

The filler function, if set, is called with the path to either a mounted directory (if contentType is filesystem) or a mapped block device (if contentType is block).

CreateFromCopy lines up with the naming on the Pool interface and takes enough details to create a volume as a copy of another one or from a snapshot of another one. It would also allow creating a new container or vm volume from an image volume.

Reply:
@stgraber ah yes, right, although did we say that we could detect the contentType from the path by detecting if it is a block device and if not assuming it is a mounted path (could check that it is a directory too).

As for CreateVolumeFromCopy, if depending on the outcome of the discussions above regarding VolumeTypes for snapshots, we would not need the snapshot boolean here. Then we would just call this function multiple times from lxdBackend based on how many snapshots (and the original volume itself) there are.

Reply:
@stgraber I've found that we can actually use the volume name itself to indicate if its a snapshot or not, so no need for additional types or arguments.

Reply:
You're going to need `srcType` in the case where you're creating a new volume of type container or virtual-machine from a volume of type image.

Reply:
Our discussion on Monday was about the arguments passed to `filler` which indeed do not need the contentType passed through to it.

We do need `contentType` for Create itself, as otherwise you have no way to indicate whether you're creating an image volume for a container (filesystem) or a VM (block).

Reply:
@stgraber I see, is contentType only needed because the "image" volume type doesn't indicate whether it is a VM image or a container image? Is it worth perhaps instead adding a concept of a VM image and then pass this via the VolumeType arg. As my understanding of this now is that we could infer the volume type if block for VM VolumeTypes and FS for containers, but we cannot at this time infer block/FS for the Image VolumeType.

Reply:
@stgraber another question, is it possible to do cross-pool copies, ie copy a volume on one pool to another pool of a different type? If so, in this case, when should the source volume be mounted (assuming it is of type that requires a per-volume mount, such as LVM and ZFS)? I've been looking in the source now and I cannot find the logic that handles mounting source volumes.

Reply:
@stgraber as discussed, my understanding is that whilst cross-pool copies are possible it should not be done via CreateVolumeFromCopy, but instead using the `MigrateCustomVolume` and `CreateCustomVolumeFromMigration` functions.